primitives/npos-elections/src/pjr.rs

```diff
@@ -0,0 +1,523 @@
+	// below, a complexity of `O(s*c)`, where `s == supports.len()` and `c == all_candidates.len()`.
+	// For large lists, that's pretty bad.
+	//
+	// A `SupportMap`, as a `BTreeMap`, has access timing of `O(lg n)`. This means that constructing
+	// the map and then indexing from it gives us timing of `O((s + c) * lg(s))`. If in the future
+	// we get access to a deterministic `HashMap`, we can further improve that to `O(s+c)`.
+	//
+	// However, it does mean allocating sufficient space to store all the data again.
+	let supports: SupportMap<AccountId> = supports.iter().cloned().collect();
+
+	let candidates = all_candidates.into_iter().enumerate().map(|(i, c)| {
+		candidates_index.insert(c.clone(), i);
+
+		// set the backing value and elected flag if the candidate is among the winners.
+		let who = c;
+		let maybe_support = supports.get(&who);
+		let elected = maybe_support.is_some();
+		let backed_stake = maybe_support.map(|support| support.total).unwrap_or_default();
+
+		Candidate { who, elected, backed_stake, ..Default::default() }.to_ptr()
+	}).collect::<Vec<_>>();
+
+	// collect all voters into a unified Vec<Voters>.
+	let voters = all_voters.into_iter().map(|(v, w, ts)| {
+		let mut edges: Vec<Edge<AccountId>> = Vec::with_capacity(ts.len());
+		for t in ts {
+			if edges.iter().any(|e| e.who == t) {
+				// duplicate edge.
+				continue;
+			}
+
+			if let Some(idx) = candidates_index.get(&t) {
+				// if this edge is among the assignments, set the weight as well.
+				let weight = assignment_map
+					.get(&v)
+					.and_then(|d| d.iter().find_map(|(x, y)| if x == &t { Some(y) } else { None }))
+					.cloned()
+					.unwrap_or_default();
+				edges.push(Edge {
+					who: t,
+					candidate: Rc::clone(&candidates[*idx]),
+					weight,
+					..Default::default()
+				});
+			}
+		}
+
+		let who = v;
+		let budget: ExtendedBalance = w.into();
+		Voter { who, budget, edges, ..Default::default() }
+	}).collect::<Vec<_>>();
+
+	(candidates, voters)
+}
+
+/// Check a solution to be t-PJR.
+///
+/// ### Semantics
+///
+/// The t-PJR property is defined in the paper ["Validator Election in Nominated Proof-of-Stake"][NPoS],
+/// section 5, definition 1.
+///
+/// In plain language, the t-PJR condition is: if there is a group of `N` voters
+/// who have `r` common candidates and can afford to support each of them with backing stake `t`
+/// (i.e `sum(stake(v) for v in voters) == r * t`), then this committee needs to be represented by at
+/// least `r` elected candidates.
+///
+/// Section 5 of the NPoS paper shows that this property can be tested by: for a feasible solution, if
+/// `Max {score(c)} < t` where c is every unelected candidate, then this solution is t-PJR. However,
+/// this test is incomplete: while every solution which passes this test satisfies t-PJR, not every
+/// solution which is t-PJR passes this test. We therefore look to a more accurate test.
```

Comment:
```suggestion
/// `Max {score(c)} < t` where c is every unelected candidate, then this solution is t-PJR.
```

I think this whole sentence is misleading and either need to be reworked or removed. I know that you are trying to address the accuracy issues. But you are phrasing it in a way as if somehow, sometimes, a PJR solution is suddenly no longer PJR. A better assumption is: If a solution `S` is PJR and is no longer PJR due to accuracy issues, it is simply because it is no longer `S` but rather `S'`, a variant thereof.

Reply:
So, that was in response to https://github.com/paritytech/substrate/pull/8160#discussion_r581943430. I think the actual intent is more like, "There exist solutions which do not pass this test but are nevertheless t-PJR." 

I'll give this some thought and try to clarify.

Reply:
What do you think of the phrasing introduced in eb943752a?

Reply:
Better, but I think it is important to clarify why: accuracy errors.

Reply:
No, I think it's possible for solutions to fail (in edge cases) for reasons that don't derive from accuracy errors. I'm basing that on this text:

> The test implies the condition, but not every solution that satifies the condition passes the test.